velox/functions/sparksql/Hash.cpp

```diff
@@ -26,21 +26,273 @@ namespace {
 
 const int32_t kDefaultSeed = 42;
 
+struct Murmur3Hash;
+struct XxHash64;
+
+// A template struct that contains the seed and return type of the hash
+// function.
+template <typename HashClass>
+struct HashTraits {};
+
+template <>
+struct HashTraits<Murmur3Hash> {
+  using SeedType = int32_t;
+  using ReturnType = int32_t;
+};
+
+template <>
+struct HashTraits<XxHash64> {
+  using SeedType = int64_t;
+  using ReturnType = int64_t;
+};
+
+// Computes the hash value of input using the hash function in HashClass.
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(int32_t input, SeedType seed) {
+  return HashClass::hashInt32(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(int64_t input, SeedType seed) {
+  return HashClass::hashInt64(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(float input, SeedType seed) {
+  return HashClass::hashFloat(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(double input, SeedType seed) {
+  return HashClass::hashDouble(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(int128_t input, SeedType seed) {
+  return HashClass::hashLongDecimal(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(Timestamp input, SeedType seed) {
+  return HashClass::hashTimestamp(input, seed);
+}
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+ReturnType hashOne(StringView input, SeedType seed) {
+  return HashClass::hashBytes(input, seed);
+}
+
+// Class to compute hashes identical to one produced by Spark.
+// Hashes are computed using the algorithm implemented in HashClass.
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class SparkVectorHasher {
+ public:
+  SparkVectorHasher(DecodedVector& decoded) : decoded_(decoded) {}
+
+  virtual ~SparkVectorHasher() = default;
+
+  // Compute the hash value of input vector at index.
+  ReturnType hashAt(vector_size_t index, SeedType seed) {
+    if (decoded_.isNullAt(index)) {
+      return seed;
+    }
+    return hashNotNullAt(index, seed);
```

Comment:
does this mean we will have a virtual function call per row when processing the hash function? Can we avoid it?

Reply:
thanks